python/rainbow/simulators/proximal_contact/blocking.py

```diff
@@ -0,0 +1,142 @@
+import numpy as np
+import numba as nb
+import networkx as nx
+from collections import defaultdict
+
+
+@nb.njit(parallel=True, nogil=True, cache=True)
+def _get_edges(J):
+    """ Get the edges of the contact graph.
+
+    Args:
+        J (ArrayLike): The contact jacobi matrix.
+
+    Returns:
+        (ArrayLike, ArrayLike): The source and target of the edges.
+    """
+    K = J.shape[0] // 4 # The number of contact points, here the each contact point has 4 rows of J.
+    step_size = 24 # Here the step_size is 24 , because each contact point has 8 non-zero blocks and each block has 3 columns.
+    cols = J.shape[1] # The number of columns of J.
+    
+    # Preallocate arrays for source and target of edges
+    max_possible_edges = (cols // step_size) * K * K
+    source = np.full(max_possible_edges, -1, dtype=np.int64)
+    target = np.full(max_possible_edges, -1, dtype=np.int64)
+    edge_count = 0
```

Comment:
This is quadratic memory usage. It will be painful for larger configurations.

Can a sparse solution not be obtained?

Reply:
Here `K` is the number of contact points, not the number of vertices. I will try to use other methods to avoid pre-memory settings.